vm/systemSmartContracts/auction.go

```diff
@@ -0,0 +1,694 @@
+}
+
+func (s *stakingAuctionSC) selection(bids []AuctionData) [][]byte {
+	nodePrice, err := s.calculateNodePrice(bids)
+	if err != nil {
+		return nil
+	}
+
+	totalQualifyingStake := big.NewFloat(0).SetInt(calcTotalQualifyingStake(nodePrice, bids))
+
+	reservePool := make(map[string]float64)
+	toBeSelectedRandom := make(map[string]float64)
+	finalSelectedNodes := make([][]byte, 0)
+	for _, validator := range bids {
+		if validator.MaxStakePerNode.Cmp(nodePrice) < 0 {
+			continue
+		}
+
+		maxPossibleNodes := big.NewInt(0).Div(validator.TotalStakeValue, nodePrice)
+		validatorQualifyingStake := big.NewFloat(0).SetInt(validator.TotalStakeValue)
+		qualifiedNodes := maxPossibleNodes.Uint64()
+
+		if maxPossibleNodes.Uint64() > uint64(len(validator.BlsPubKeys)) {
+			validatorQualifyingStake = big.NewFloat(0).SetInt(big.NewInt(0).Mul(nodePrice, big.NewInt(int64(len(validator.BlsPubKeys)))))
+			qualifiedNodes = uint64(len(validator.BlsPubKeys))
+		}
+
+		proportionOfTotalStake := big.NewFloat(0).Quo(validatorQualifyingStake, totalQualifyingStake)
+		proportion, _ := proportionOfTotalStake.Float64()
+		allocatedNodes := float64(qualifiedNodes) * proportion
+		numAllocatedNodes := uint64(allocatedNodes)
+		if allocatedNodes-float64(numAllocatedNodes) > 0.99 {
+			numAllocatedNodes += 1
+		} else {
+			selectorProp := allocatedNodes - float64(numAllocatedNodes)
+			toBeSelectedRandom[string(validator.BlsPubKeys[numAllocatedNodes])] = selectorProp
+
+			for i := numAllocatedNodes + 1; i < uint64(len(validator.BlsPubKeys)); i++ {
+				reservePool[string(validator.BlsPubKeys[i])] = selectorProp
+			}
+		}
+
+		if numAllocatedNodes > 0 {
+			finalSelectedNodes = append(finalSelectedNodes, validator.BlsPubKeys[:numAllocatedNodes]...)
+		}
+	}
+
+	config := s.getConfig(s.eei.BlockChainHook().CurrentEpoch())
+	stillNeeded := uint32(0)
+	if config.NumNodes > uint32(len(finalSelectedNodes)) {
+		stillNeeded = config.NumNodes - uint32(len(finalSelectedNodes))
+	}
+
+	randomlySelected := s.selectRandomly(toBeSelectedRandom, stillNeeded)
+	finalSelectedNodes = append(finalSelectedNodes, randomlySelected...)
+
+	if config.NumNodes > uint32(len(finalSelectedNodes)) {
+		stillNeeded = config.NumNodes - uint32(len(finalSelectedNodes))
+		randomlySelected := s.selectRandomly(reservePool, stillNeeded)
+		finalSelectedNodes = append(finalSelectedNodes, randomlySelected...)
+	}
+
+	return finalSelectedNodes
+}
+
+func (s *stakingAuctionSC) selectRandomly(selectable map[string]float64, numNeeded uint32) [][]byte {
+	randomlySelected := make([][]byte, 0)
+	if numNeeded == 0 {
+		return randomlySelected
+	}
+
+	expandedList := make([]string, 0)
+	for key, proportion := range selectable {
+		expandVal := uint8(proportion)*10 + 1
+		for i := uint8(0); i < expandVal; i++ {
+			expandedList = append(expandedList, key)
+		}
+	}
+
+	random := s.eei.BlockChainHook().CurrentRandomSeed()
+	randomSeed := big.NewInt(0).SetBytes(random[:8])
+	rand.Seed(randomSeed.Int64())
```

Comment:
Relying on the rand library like this might lead to unpredictable results if there is a concurrent rand call. 
Strong suggestion to use fisherYatesShuffle implementation (as can be seen dataRetriever/resolvers/topicResolverSender package) in which you have to provide a randomizer. Randomizer should be injected on the stakingAuctionSC as to be able to better test this functionality.